modules/dcache/src/main/java/org/dcache/services/ssh2/Ssh2Admin.java

```diff
@@ -318,6 +327,66 @@ public boolean authenticate(String userName, PublicKey key,
         }
     }
 
+    private class RemoteHostValidator {
+
+        private boolean isValidHost(String line, ServerSession session) {
+            for (String linePart : line.split(" ")) {
+                if (linePart.startsWith("from=")) {
+                    for (String pattern : strip(linePart).split(",")){
+                        boolean patternIsNegated = pattern.startsWith("!");
+                        boolean patternMatchesHost = patternMatchesHost(pattern, session);
+                        if (patternIsNegated==patternMatchesHost){
+                            return false;
+                        }
+                    }
+                }
+            }
+            return true;
+        }
+
+        private boolean patternMatchesHost(String pattern, ServerSession session){
+            SocketAddress remote = session.getClientAddress();
+            if (remote instanceof InetSocketAddress) {
+                String remoteAddress = ((InetSocketAddress) remote).getAddress().getHostAddress();
+                String remoteName = ((InetSocketAddress) remote).getHostName();
+                Pattern p = Pattern.compile(convertPatternToRegex(pattern));
+                Matcher addressMatcher = p.matcher(remoteAddress);
+                Matcher nameMatcher = p.matcher(remoteName);
+                return (addressMatcher.matches()
+                        || nameMatcher.matches()
+                        || addressInCidrRange(pattern, remoteAddress));
+            }
+            return false;
+        }
+
+        private boolean addressInCidrRange(String cidrAddress, String remoteAddress) {
+            cidrAddress = cidrAddress.replace("!","");
```

Comment:
This code, to remove of the `!`, appears both here and convertPatternToRegex, suggesting it should be handled in a common location (e.g., `patternMatchesHost`).  Also, since you know it must be the first character, removing it can be done much simpler (and faster): just create a substring starting from the second character.